Implement async quotes scraper

vladyslav-tmf opened the pull request:
- Implemented asynchronous scraping with BeautifulSoup.
- Implemented CSV export functionality.
- Added logging to both console and file.
- Added semaphore to limit concurrent requests.

vladyslav-tmf committed:
Implement quotes scraper

- Implemented page-by-page scraping with BeautifulSoup.
- Implemented CSV export functionality.
- Added logging to both console and file.
- Updated .flake8 to exclude all virtual environments.
- Modified .gitignore to include .csv and .log files.
- Added beautifulsoup4 and requests dependencies to requirements.txt.

vladyslav-tmf committed:
Refactor scraper to use aiohttp for asynchronous requests

- Replaced requests with aiohttp for async HTTP requests.
- Implemented parallel page processing with asyncio.
- Added semaphore to limit concurrent requests.

mateacademy-ai-mentor commented on app/parse.py:
```diff
@@ -8,8 +19,127 @@ class Quote:
+    return [parse_single_quote(quote_div) for quote_div in quotes_divs]
+
+
+def get_next_page_url(page_soup: Tag) -> str | None:
```
The type hint `str | None` is not valid in Python versions below 3.10. If you are using an earlier version, consider using `Optional[str]` from the `typing` module instead.

mateacademy-ai-mentor commented on app/parse.py:
```diff
@@ -8,8 +19,127 @@
+    url: str,
+    page_number: int,
+    semaphore: asyncio.Semaphore,
+) -> tuple[list[Quote], str | None]:
```
Similar to the previous comment, the type hint `str | None` may not be compatible with Python versions below 3.10. Use `Optional[str]` if necessary.

mateacademy-ai-mentor reviewed (approved):
Great job on implementing the asynchronous web scraper! 🎉 Your code is well-structured and meets all the task requirements effectively. While there are some minor improvements suggested, such as using `Optional[str]` for type hints to ensure compatibility with Python versions below 3.10, these do not impact the core functionality. Keep up the good work, and consider these suggestions for future projects to enhance compatibility and maintainability. 😊

✨ Please rate this review in the Platform's notification center. I will be grateful for your feedback! ✨

vladyslav-tmf committed:
Enhance quote scraper to include author biographies

- Created Author dataclass to store author information.
- Implemented function to extract author URLs from quotes.
- Developed function to parse author biographies from their pages.
- Updated process_single_page to collect author data alongside quotes.
- Modified get_all_quotes to return both quotes and authors.
- Introduced write_authors_to_csv function to save author data to a CSV file.
- Updated logging to track author data collection.
- Updated async_main and main functions to handle authors' CSV export.

Arsen-hrynevych commented on app/parse.py:
```diff
@@ -8,9 +28,185 @@ class Quote:
+        page_number = 2
+        current_url = next_url
+
+        while current_url:
```
Instead of manually iterating through pages sequentially for page_urls, you can fetch them concurrently. This eliminates the need for the loop:

Arsen-hrynevych reviewed (changes requested):
Split your logic, for example, into a file named utils.py

vladyslav-tmf committed:
Optimize scraper performance

- Implement parallel author biography fetching
- Add execution time tracking

vladyslav-tmf committed:
Split code into modules

- Create models.py for data classes
- Move helper functions to utils.py
- Keep async logic in parse.py

Arsen-hrynevych reviewed (changes requested):
In my opinion, you are overengineering. I suggest simplifying it. You can still split the functionality into functions and make the code more readable by using blank spaces, etc.
